src/lib/support/jsontlv/TlvToJson.cpp

```diff
@@ -0,0 +1,371 @@
+    }
+
+    void SetElementType(const char * elemType) { elementType = elemType; }
+
+    void SetArrayElementType(const char * subType) { subElementType = subType; }
+
+    void GenerateJsonKeyStr(char * buf) const
+    {
+        int len = 0;
+        len     = sprintf(buf, "%u:", tag);
+        len += sprintf(&buf[len], "%s", elementType);
+        if (strcmp(elementType, kElementTypeArray) == 0)
+        {
+            sprintf(&buf[len], "-%s", subElementType);
+        }
+    }
+
+    ContainerType containerType = kStruct;
+    unsigned int tag            = 0;
+    const char * elementType    = nullptr;
+    const char * subElementType = nullptr;
+};
+
+/*
+ * For now, let's put a bound of the maximum length of a byte/char string to be the size of an IPv6
+ * MTU. While this is smaller than that of the limit defined in the data model specification,
+ * strings by virtue of not being chunked are intrinsically limited in size to the size of the encompassing packet.
+ */
+static constexpr uint16_t kMaxStringLen = 1280;
+
+/*
+ * This templated function inserts a key/value pair into the Json value object.
+ * The value is templated to be of type T and accepts any of the following primitive
+ * types:
+ *      bool, uint*_t, int*_t, char *, float, double.
+ *
+ * This method uses the provided key context to deduce the type of element being added.
+ *
+ */
+template <typename T>
+void InsertKeyValue(Json::Value & json, const KeyContext & keyContext, T val)
+{
+    if (json.isArray())
+    {
+        json.append(val);
+    }
+    else
+    {
+        // This needs to accommodate the largest key value 'FieldId:ElementType-SubElementType'.
+        // The largest 32-bit integer FieldId represented as a string is 11 characters long.
+        // The largest ElementType and SubElementType strings are 6 characters long each.
+        // Tack on 3 bytes for the ':' character, '-' character, and null terminator.
+        char keyBuf[26];
+        keyContext.GenerateJsonKeyStr(keyBuf);
+        json[keyBuf] = val;
+    }
+}
+
+static CHIP_ERROR TlvToJson(TLV::TLVReader & reader, KeyContext context, Json::Value & jsonObj);
+
+/*
+ * Given a TLVReader positioned at TLV structure this function:
+ *   - enters structure
+ *   - converts all elements of a structure into JSON object representation
+ *   - exits structure
+ */
+CHIP_ERROR TlvStructToJson(TLV::TLVReader & reader, Json::Value & jsonObj)
+{
+    CHIP_ERROR err;
+    TLV::TLVType containerType;
+
+    ReturnErrorOnFailure(reader.EnterContainer(containerType));
+
+    while ((err = reader.Next()) == CHIP_NO_ERROR)
+    {
+        TLV::Tag tag = reader.GetTag();
+        VerifyOrReturnError(TLV::IsContextTag(tag) ||
+                                (TLV::IsProfileTag(tag) && (TLV::ProfileIdFromTag(tag) == TLV::kCommonProfileId)),
+                            CHIP_ERROR_INVALID_TLV_TAG);
+        KeyContext context(static_cast<chip::FieldId>(TLV::TagNumFromTag(tag)));
```

Comment:
What's the point of supporting common profile tags only if the tag number is > 255?  

I'm really struggling to understand the goals of the common profile support here:

1. Nothing in the spec uses profile tags right now, common or otherwise.
2. The restriction to > 255 is odd.

What use cases is this common profile support actually aiming for?

Reply:
@emargolis Please see above.

Reply:
The support for Common Profile Tags was introduced here to be able to encode attribute id as a Common Profile Tag Number and the associated data as a TLV structure. This was necessary for some of our use cases.

I would like to ask @mrjerryjohns @turon @yunhanw-google to comment here.

Is it OK to limit Common Profile Tag Number to always be higher than 255 or we need to update the format to be able to distinguish between Context and Common Profile tags. 

For example the Common Profile tag value can be encoded with a `CP` prefix in the Json name:
```
{
    "name:CP1:INT" : 42
}
```

Reply:
Follow up: https://github.com/project-chip/connectedhomeip/issues/28387

Reply:
> to be able to encode attribute id as a Common Profile Tag Number

Attribute IDs are almost all smaller than 255... So unless they were going to be encoded with some offset.

I keep coming back to the basic question: can we please clearly describe which TLVs this is trying to represent in JSON?  Because _nowhere_ do we have attribute IDs as tags in TLVs right now, in the spec.  And the set of TLVs we are trying to represent affects how we go about doing things....

It's really hard to review code for correctness when people keep not telling you what it's meant to be trying to do.  @andy31415 @robszewczyk

Reply:
I think it seems reasonable to do one of:
  - have to support full spec TLV (which means some things not used in the spec could be skipped and errored out on)
  - have to support full TLV ... which means proper encode/decode of the profile tags
  
The latter seems more generic, so we should probably go with that.

Reply:
Had some more discussions around this, so my understanding of where these tags come from is from a desire to compact encoding of `*PathIB` identifiers. The intent was to have human-friendly attribute representation of `{"123:BOOL": true}` represent `Boolean attribute 123 has value of true`.

As a sideffect, when commandids/eventids/attributeids are into JSON, the PR tries to keep allowing these JSON-TLV conversion for such 32-bit numbers without losing their value (but not caring about actual TLV encoding).


So current TLV json wants to:
  - fully support bi-directional data model payloads encoding/decoding
  - support preserving 32-bit values of tag values (but not their type) specifically for potential attribute/command/event id encoding
  
  The 32-bit tag preserving is what has the rule of:
     - use context tag for 0-255 (because this is compatible with spec encoding of struct things)
     - use profile tag for anything larger, because we have no context tag defined in the spec for large numbers
     
 This seems to imply that json will know the valid ranges of things, so I believe we should be able to error out if we encouter values that are not expected (e.g. profile tag that is "too small").

Reply:
@andy31415 

This is an interesting twist to use "implicit profile id" instead of common profile tag. I would suggest to go thurder and encode profile id information in the Json name. Specifically, the "profile id" can be combined with the "tag number" to be encoded as a 64-bit positive integer:
`field_id = static_cast<uint64_t>(ProfileIdFromTag(tag)) << 32 || TagNumFromTag(tag)`

For example if profile id is `0xFF01` and tag number is `0x1` then the boolean element will be encoded as:
```
{
    "name:280379760050177:BOOL" : true
}
```
where `0xFF01_0000_0001 = 280379760050177`

We can maybe go further and make it more user friendly and require Profile tags to be represented in HEX format, for example:
```
{
    "name:0xFF0100000001:BOOL" : true
}
```
If we require profile tags to be encoded in HEX format with `0x` prefix and Context tags as decimals then we won't have collision issue of the common profile tag numbers smaller than 256.